httpbp/headers.go

```diff
@@ -0,0 +1,209 @@
+	// HTTP request headers.
+	SpanSampledHeader = "X-Sampled"
+
+	// TraceIDHeader is the key use to get the trace ID from the HTTP
+	// request headers.
+	TraceIDHeader = "X-Trace"
+)
+
+type headerContextKey int
+
+const (
+	edgeContextContextKey headerContextKey = iota
+	traceIDContextKey
+	parentIDContextKey
+	spanIDContextKey
+	spanFlagsContextKey
+	spanSampledContextKey
+)
+
+// HeaderContextKey is an Enum used to get HTTP headers from a context object.
+//
+// It is not used as the actual key in the context, these enums are mapped to
+// un-exported values that serve as the actual context keys.
+type HeaderContextKey int
+
+const (
+	// EdgeContextContextKey is the key for the raw edge request context
+	EdgeContextContextKey HeaderContextKey = iota
+
+	// TraceIDContextKey is the header for the trace ID passed by the caller
+	TraceIDContextKey
+
+	// ParentIDContextKey is the header for the parent ID passed by the caller
+	ParentIDContextKey
+
+	// SpanIDContextKey is the header for the span ID passed by the caller
+	SpanIDContextKey
+
+	// SpanFlagsContextKey is the header for the span flags passed by the caller
+	SpanFlagsContextKey
+
+	// SpanSampledContextKey is the header for the sampled flag passed by the caller
+	SpanSampledContextKey
+)
+
+var headerKeyToContextKey = map[HeaderContextKey]headerContextKey{
+	EdgeContextContextKey: edgeContextContextKey,
+	TraceIDContextKey:     traceIDContextKey,
+	ParentIDContextKey:    parentIDContextKey,
+	SpanIDContextKey:      spanIDContextKey,
+	SpanFlagsContextKey:   spanFlagsContextKey,
+	SpanSampledContextKey: spanSampledContextKey,
+}
+
+// GetHeader returns the HTTP header stored on the context at key.
+func GetHeader(ctx context.Context, key HeaderContextKey) (header string) {
+	if contextKey, ok := headerKeyToContextKey[key]; ok {
+		if h, ok := ctx.Value(contextKey).(string); ok {
+			header = h
+		}
+	}
+	return
+}
+
+// Headers is an interface to collect all of the HTTP headers for a particular
+// baseplate resource (spans and edge contexts) into a struct that provides an
+// easy way to convert them into HTTP headers.
+//
+// This interface exists so we can avoid having to do runtime checks on maps to
+// ensure that they have the right keys set when we are trying to sign or verify
+// a set of HTTP headers.
+type Headers interface {
+	// AsMap returns the Headers struct as a map of header keys to header
+	// values.
+	AsMap() map[string]string
+}
+
+// EdgeContextHeaders implements the Headers interface for HTTP EdgeContext
+// headers.
+type EdgeContextHeaders struct {
+	EdgeRequest string
+}
+
+// NewEdgeContextHeaders returns a new EdgeContextHeaders object from the given
+// HTTP headers.
+func NewEdgeContextHeaders(h http.Header) *EdgeContextHeaders {
+	return &EdgeContextHeaders{
+		EdgeRequest: h.Get(EdgeContextHeader),
+	}
+}
+
+// AsMap returns the EdgeContextHeaders as a map of header keys to header
+// values.
+func (s *EdgeContextHeaders) AsMap() map[string]string {
```

Comment:
nit: This does not need to be a pointer receiver. Same for `SpanHeaders`

Reply:
Is this because strings aren't actually copied?  Since they're immutable you can just copy the pointer and metadata?

Reply:
Yes.